### ISSUE TITLE
loss =87.3

### ISSUE THREAD

**Issue by double-vane** [wy1iu/sphereface#40]
Hi, when I trained the sphereface-27 with my dataset, I always run into a problem：loss=87.33, and my prototxt as shown below:

`layer {
  name: "fc6-"
  type: "MarginInnerProduct"
  bottom: "fc5"
  bottom: "label"
  top: "fc6"
  top: "lambda"
  param {
    lr_mult: 1
    decay_mult: 1
  }
  margin_inner_product_param {
    num_output: 11954
    type: QUADRUPLE
    weight_filler {
      type: "xavier"
    }
    base: 10
    gamma: 2e-05
    power: 1
    iteration: 0
    lambda_min: 5
  }
}
`
`I0922 23:43:59.428972 192876 solver.cpp:219] Iteration 7000 (0.325212 iter/s, 307.492s/100 iters), loss = 10.0837
I0922 23:43:59.429009 192876 solver.cpp:238]     Train net output #0: accuracy = 0.574219
I0922 23:43:59.429015 192876 solver.cpp:238]     Train net output #1: lambda = 8.77178
I0922 23:43:59.429021 192876 solver.cpp:238]     Train net output #2: softmax_loss = 10.0837 (* 1 = 10.0837 loss)
I0922 23:43:59.429025 192876 sgd_solver.cpp:105] Iteration 7000, lr = 1e-05
I0922 23:45:04.017473 192883 data_layer.cpp:73] Restarting data prefetching from start.
I0922 23:49:06.527763 192876 solver.cpp:219] Iteration 7100 (0.325632 iter/s, 307.096s/100 iters), loss = 7.03009
I0922 23:49:06.527886 192876 solver.cpp:238]     Train net output #0: accuracy = 0.6875
I0922 23:49:06.527894 192876 solver.cpp:238]     Train net output #1: lambda = 8.75641
I0922 23:49:06.527901 192876 solver.cpp:238]     Train net output #2: softmax_loss = 7.03009 (* 1 = 7.03009 loss)
I0922 23:49:06.527905 192876 sgd_solver.cpp:105] Iteration 7100, lr = 1e-05
I0922 23:54:13.578683 192876 solver.cpp:219] Iteration 7200 (0.325682 iter/s, 307.048s/100 iters), loss = 8.27169
I0922 23:54:13.578783 192876 solver.cpp:238]     Train net output #0: accuracy = 0.65625
I0922 23:54:13.578791 192876 solver.cpp:238]     Train net output #1: lambda = 8.74111
I0922 23:54:13.578797 192876 solver.cpp:238]     Train net output #2: softmax_loss = 8.27169 (* 1 = 8.27169 loss)
I0922 23:54:13.578801 192876 sgd_solver.cpp:105] Iteration 7200, lr = 1e-05
I0922 23:59:20.621175 192876 solver.cpp:219] Iteration 7300 (0.325691 iter/s, 307.039s/100 iters), loss = 12.1695
I0922 23:59:20.621300 192876 solver.cpp:238]     Train net output #0: accuracy = 0.617188
I0922 23:59:20.621309 192876 solver.cpp:238]     Train net output #1: lambda = 8.72585
I0922 23:59:20.621316 192876 solver.cpp:238]     Train net output #2: softmax_loss = 12.1695 (* 1 = 12.1695 loss)
I0922 23:59:20.621321 192876 sgd_solver.cpp:105] Iteration 7300, lr = 1e-05
I0923 00:04:27.686278 192876 solver.cpp:219] Iteration 7400 (0.325667 iter/s, 307.062s/100 iters), loss = 10.945
I0923 00:04:27.686388 192876 solver.cpp:238]     Train net output #0: accuracy = 0.664062
I0923 00:04:27.686394 192876 solver.cpp:238]     Train net output #1: lambda = 8.71065
I0923 00:04:27.686416 192876 solver.cpp:238]     Train net output #2: softmax_loss = 10.945 (* 1 = 10.945 loss)
I0923 00:04:27.686421 192876 sgd_solver.cpp:105] Iteration 7400, lr = 1e-05
I0923 00:09:29.417439 192876 solver.cpp:448] Snapshotting to binary proto file /home/unique/data-face/face_model/WT_sphereface27-40_iter_7500.caffemodel
I0923 00:09:31.971767 192876 sgd_solver.cpp:273] Snapshotting solver state to binary proto file /home/unique/data-face/face_model/WT_sphereface27-40_iter_7500.solverstate
I0923 00:09:35.155453 192876 solver.cpp:219] Iteration 7500 (0.325239 iter/s, 307.466s/100 iters), loss = 10.7305
I0923 00:09:35.155522 192876 solver.cpp:238]     Train net output #0: accuracy = 0.644531
I0923 00:09:35.155529 192876 solver.cpp:238]     Train net output #1: lambda = 8.6955
I0923 00:09:35.155535 192876 solver.cpp:238]     Train net output #2: softmax_loss = 10.7305 (* 1 = 10.7305 loss)
I0923 00:09:35.155540 192876 sgd_solver.cpp:105] Iteration 7500, lr = 1e-05
I0923 00:14:42.234946 192876 solver.cpp:219] Iteration 7600 (0.325652 iter/s, 307.076s/100 iters), loss = 12.6641
I0923 00:14:42.235061 192876 solver.cpp:238]     Train net output #0: accuracy = 0.652344
I0923 00:14:42.235069 192876 solver.cpp:238]     Train net output #1: lambda = 8.6804
I0923 00:14:42.235076 192876 solver.cpp:238]     Train net output #2: softmax_loss = 12.6641 (* 1 = 12.6641 loss)
I0923 00:14:42.235080 192876 sgd_solver.cpp:105] Iteration 7600, lr = 1e-05
I0923 00:19:49.294009 192876 solver.cpp:219] Iteration 7700 (0.325674 iter/s, 307.056s/100 iters), loss = 13.261
I0923 00:19:49.294123 192876 solver.cpp:238]     Train net output #0: accuracy = 0.617188
I0923 00:19:49.294131 192876 solver.cpp:238]     Train net output #1: lambda = 8.66536
I0923 00:19:49.294138 192876 solver.cpp:238]     Train net output #2: softmax_loss = 13.261 (* 1 = 13.261 loss)
I0923 00:19:49.294142 192876 sgd_solver.cpp:105] Iteration 7700, lr = 1e-05
I0923 00:24:56.366257 192876 solver.cpp:219] Iteration 7800 (0.32566 iter/s, 307.069s/100 iters), loss = 13.8046
I0923 00:24:56.366369 192876 solver.cpp:238]     Train net output #0: accuracy = 0.628906
I0923 00:24:56.366377 192876 solver.cpp:238]     Train net output #1: lambda = 8.65037
I0923 00:24:56.366386 192876 solver.cpp:238]     Train net output #2: softmax_loss = 13.8046 (* 1 = 13.8046 loss)
I0923 00:24:56.366391 192876 sgd_solver.cpp:105] Iteration 7800, lr = 1e-05
I0923 00:30:03.426719 192876 solver.cpp:219] Iteration 7900 (0.325672 iter/s, 307.057s/100 iters), loss = 14.3722
I0923 00:30:03.426854 192876 solver.cpp:238]     Train net output #0: accuracy = 0.640625
I0923 00:30:03.426863 192876 solver.cpp:238]     Train net output #1: lambda = 8.63543
I0923 00:30:03.426870 192876 solver.cpp:238]     Train net output #2: softmax_loss = 14.3722 (* 1 = 14.3722 loss)
I0923 00:30:03.426877 192876 sgd_solver.cpp:105] Iteration 7900, lr = 1e-05
I0923 00:35:05.096544 192876 solver.cpp:448] Snapshotting to binary proto file /home/unique/data-face/face_model/WT_sphereface27-40_iter_8000.caffemodel
I0923 00:35:07.648367 192876 sgd_solver.cpp:273] Snapshotting solver state to binary proto file /home/unique/data-face/face_model/WT_sphereface27-40_iter_8000.solverstate
I0923 00:35:10.833468 192876 solver.cpp:219] Iteration 8000 (0.325305 iter/s, 307.403s/100 iters), loss = 17.3922
I0923 00:35:10.833523 192876 solver.cpp:238]     Train net output #0: accuracy = 0.648438
I0923 00:35:10.833528 192876 solver.cpp:238]     Train net output #1: lambda = 8.62054
I0923 00:35:10.833534 192876 solver.cpp:238]     Train net output #2: softmax_loss = 17.3922 (* 1 = 17.3922 loss)
I0923 00:35:10.833539 192876 sgd_solver.cpp:105] Iteration 8000, lr = 1e-05
I0923 00:40:17.756762 192876 solver.cpp:219] Iteration 8100 (0.325818 iter/s, 306.92s/100 iters), loss = 17.3765
I0923 00:40:17.756888 192876 solver.cpp:238]     Train net output #0: accuracy = 0.644531
I0923 00:40:17.756896 192876 solver.cpp:238]     Train net output #1: lambda = 8.6057
I0923 00:40:17.756903 192876 solver.cpp:238]     Train net output #2: softmax_loss = 17.3765 (* 1 = 17.3765 loss)
I0923 00:40:17.756908 192876 sgd_solver.cpp:105] Iteration 8100, lr = 1e-05
I0923 00:45:24.701261 192876 solver.cpp:219] Iteration 8200 (0.325795 iter/s, 306.941s/100 iters), loss = 17.4908
I0923 00:45:24.701335 192876 solver.cpp:238]     Train net output #0: accuracy = 0.625
I0923 00:45:24.701341 192876 solver.cpp:238]     Train net output #1: lambda = 8.59092
I0923 00:45:24.701349 192876 solver.cpp:238]     Train net output #2: softmax_loss = 17.4908 (* 1 = 17.4908 loss)
I0923 00:45:24.701354 192876 sgd_solver.cpp:105] Iteration 8200, lr = 1e-05
I0923 00:50:31.609172 192876 solver.cpp:219] Iteration 8300 (0.325834 iter/s, 306.905s/100 iters), loss = 21.0124
I0923 00:50:31.609232 192876 solver.cpp:238]     Train net output #0: accuracy = 0.625
I0923 00:50:31.609238 192876 solver.cpp:238]     Train net output #1: lambda = 8.57618
I0923 00:50:31.609246 192876 solver.cpp:238]     Train net output #2: softmax_loss = 21.0124 (* 1 = 21.0124 loss)
I0923 00:50:31.609249 192876 sgd_solver.cpp:105] Iteration 8300, lr = 1e-05
I0923 00:55:38.455067 192876 solver.cpp:219] Iteration 8400 (0.3259 iter/s, 306.843s/100 iters), loss = 26.4586
I0923 00:55:38.455188 192876 solver.cpp:238]     Train net output #0: accuracy = 0.546875
I0923 00:55:38.455196 192876 solver.cpp:238]     Train net output #1: lambda = 8.5615
I0923 00:55:38.455205 192876 solver.cpp:238]     Train net output #2: softmax_loss = 26.4586 (* 1 = 26.4586 loss)
I0923 00:55:38.455210 192876 sgd_solver.cpp:105] Iteration 8400, lr = 1e-05
I0923 01:00:39.942590 192876 solver.cpp:448] Snapshotting to binary proto file /home/unique/data-face/face_model/WT_sphereface27-40_iter_8500.caffemodel
I0923 01:00:42.497731 192876 sgd_solver.cpp:273] Snapshotting solver state to binary proto file /home/unique/data-face/face_model/WT_sphereface27-40_iter_8500.solverstate
I0923 01:00:45.680984 192876 solver.cpp:219] Iteration 8500 (0.325497 iter/s, 307.223s/100 iters), loss = 27.8068
I0923 01:00:45.681020 192876 solver.cpp:238]     Train net output #0: accuracy = 0.578125
I0923 01:00:45.681025 192876 solver.cpp:238]     Train net output #1: lambda = 8.54686
I0923 01:00:45.681032 192876 solver.cpp:238]     Train net output #2: softmax_loss = 27.8068 (* 1 = 27.8068 loss)
I0923 01:00:45.681036 192876 sgd_solver.cpp:105] Iteration 8500, lr = 1e-05
I0923 01:05:52.483103 192876 solver.cpp:219] Iteration 8600 (0.325946 iter/s, 306.799s/100 iters), loss = 28.9902
I0923 01:05:52.483191 192876 solver.cpp:238]     Train net output #0: accuracy = 0.574219
I0923 01:05:52.483199 192876 solver.cpp:238]     Train net output #1: lambda = 8.53228
I0923 01:05:52.483207 192876 solver.cpp:238]     Train net output #2: softmax_loss = 28.9902 (* 1 = 28.9902 loss)
I0923 01:05:52.483211 192876 sgd_solver.cpp:105] Iteration 8600, lr = 1e-05
I0923 01:10:59.224231 192876 solver.cpp:219] Iteration 8700 (0.326011 iter/s, 306.738s/100 iters), loss = 45.9285
I0923 01:10:59.224365 192876 solver.cpp:238]     Train net output #0: accuracy = 0.429688
I0923 01:10:59.224373 192876 solver.cpp:238]     Train net output #1: lambda = 8.51774
I0923 01:10:59.224380 192876 solver.cpp:238]     Train net output #2: softmax_loss = 45.9285 (* 1 = 45.9285 loss)
I0923 01:10:59.224385 192876 sgd_solver.cpp:105] Iteration 8700, lr = 1e-05
I0923 01:16:05.965823 192876 solver.cpp:219] Iteration 8800 (0.326011 iter/s, 306.738s/100 iters), loss = 62.0409
I0923 01:16:05.965976 192876 solver.cpp:238]     Train net output #0: accuracy = 0.289062
I0923 01:16:05.965996 192876 solver.cpp:238]     Train net output #1: lambda = 8.50326
I0923 01:16:05.966014 192876 solver.cpp:238]     Train net output #2: softmax_loss = 62.0409 (* 1 = 62.0409 loss)
I0923 01:16:05.966025 192876 sgd_solver.cpp:105] Iteration 8800, lr = 1e-05
I0923 01:21:12.843422 192876 solver.cpp:219] Iteration 8900 (0.325866 iter/s, 306.874s/100 iters), loss = 67.5738
I0923 01:21:12.843535 192876 solver.cpp:238]     Train net output #0: accuracy = 0.222656
I0923 01:21:12.843542 192876 solver.cpp:238]     Train net output #1: lambda = 8.48882
I0923 01:21:12.843550 192876 solver.cpp:238]     Train net output #2: softmax_loss = 67.5738 (* 1 = 67.5738 loss)
I0923 01:21:12.843554 192876 sgd_solver.cpp:105] Iteration 8900, lr = 1e-05
I0923 01:26:14.555048 192876 solver.cpp:448] Snapshotting to binary proto file /home/unique/data-face/face_model/WT_sphereface27-40_iter_9000.caffemodel
I0923 01:26:17.105406 192876 sgd_solver.cpp:273] Snapshotting solver state to binary proto file /home/unique/data-face/face_model/WT_sphereface27-40_iter_9000.solverstate
I0923 01:26:20.287829 192876 solver.cpp:219] Iteration 9000 (0.325265 iter/s, 307.441s/100 iters), loss = 65.0543
I0923 01:26:20.287881 192876 solver.cpp:238]     Train net output #0: accuracy = 0.238281
I0923 01:26:20.287886 192876 solver.cpp:238]     Train net output #1: lambda = 8.47443
I0923 01:26:20.287894 192876 solver.cpp:238]     Train net output #2: softmax_loss = 65.0543 (* 1 = 65.0543 loss)
I0923 01:26:20.287899 192876 sgd_solver.cpp:105] Iteration 9000, lr = 1e-05
I0923 01:31:27.157495 192876 solver.cpp:219] Iteration 9100 (0.325875 iter/s, 306.866s/100 iters), loss = 87.3365
I0923 01:31:27.157562 192876 solver.cpp:238]     Train net output #0: accuracy = 0
I0923 01:31:27.157569 192876 solver.cpp:238]     Train net output #1: lambda = 8.46009
I0923 01:31:27.157577 192876 solver.cpp:238]     Train net output #2: softmax_loss = 87.3365 (* 1 = 87.3365 loss)
`
What should I do to solve the problem? Thanks!

**Comment by wy1iu** [comment]
see https://github.com/wy1iu/LargeMargin_Softmax_Loss#notes-for-training

**Comment by double-vane** [comment]
>  For large and difficult datasets, you should first try to set lambda_min as 5 or 10
Is this one?
But, when I use sphereface-20, I can get a small lambda_min and the results is good.

**Comment by wy1iu** [comment]
It really depends on the dataset and the network architecture. There is no universally good hyper-parameter in all cases. You should change it depending on your task, dataset and the network.

**Comment by double-vane** [comment]
Ok, How to determine lambda_min? 
train softmax_loss ? train accuracy? or test results?

**Comment by wy1iu** [comment]
Maybe trying a large value first, and gradually decreasing it could be a good strategy.

Besides the lambda_min. the other hyperparameters also affect the convergence.

**Comment by double-vane** [comment]
Okay, I down it from 1000, and my boss wants me to drop it to 5.
I'll try it for more numbers!
Thanks again.
Best!

**Comment by fromwhzz** [comment]
87.3365 is a magic number. if the data is nan, the softmaxloss will be 87.3365. You can print the debug info  to see the wrong. And I think you can print the x_norm and weight norm in the margin_inner product , maybe the x_norm is 0 which will make the data to be nan.

**Comment by chiukeung** [comment]
@fromwhzz I suffered the same problem and according to your suggestion, I found L1 norm = (nan, nan); L2 norm = (nan, nan), loss is always 87.3365. What should I do to solve it? Thanks a lot.

**Comment by chandraOloid** [comment]
@wy1iu 
 For a custom dataset, I have 850 ids for train and 877 ids for test with 909 unique ids. 

 what number should I use for the margin_inner_product in softmax loss ? Also, I have the same problem with softmaxloss=87.335 and final of net.forward with a known image is nan. 

 Thanks
 Chandra
